buildlib/azure-pipelines-perf.yml

```diff
@@ -0,0 +1,129 @@
+            clean: true
+            fetchDepth: 1
+
+          - bash: |
+              set -x
+              mkdir -p $(WorkDir)
+              mv $(Build.SourcesDirectory)/* $(WorkDir)    
+            displayName: Prepare WorkDir with code
+
+
+  - stage: Performance
+    dependsOn: Prepare
+    jobs:
+      - job: Perf
+        displayName: Performance testing
+        pool:
+          name: MLNX
+          demands:
+          - ucx_perf
+
+        steps:
+          - checkout: none
+            clean: true
+
+          - bash: |
+              printf '\n%79s\n\n' | tr ' ' =
+              echo "Triggered by repo: $(Build.Repository.Name)"
+              echo "Triggering repo URL: $(Build.Repository.Uri)"
+              echo "The branch of the triggering repo the build was queued for: $(Build.SourceBranch)"
+              echo "PR's merge message: $(Build.SourceVersionMessage)"
+              
+              # Extract commit SHAs from the PR's merge message.
+              SHA_After=$(echo $(Build.SourceVersionMessage) | awk '{print $2}')
+              SHA_Before=$(echo $(Build.SourceVersionMessage) | awk '{print $4}')
+                            
+              printf '\n%79s\n\n' | tr ' ' =
+              echo "Base SHA from Git log:       $(cd $(WorkDir)/ucx && git rev-parse HEAD^)"
+              echo "Base SHA from merge message: $SHA_Before"
+
+              printf '\n%79s\n\n' | tr ' ' =
+              echo "Latest SHA from Git log:  $(cd $(WorkDir)/ucx && git rev-parse HEAD)"
+              echo "Latest SHA from Azure:    $(Build.SourceVersion)"
+              echo "Topic commit to merge:    $SHA_After"
+              echo -e "The merge commit will be used for testing. \n"
+            displayName: Debug info
+            workingDirectory: $(WorkDir)
+
+          - bash: |
+              set -eE
+              # Extract commit SHAs from the PR's merge message.
+              # SHA_After=$(Build.SourceVersion)
+              SHA_After=$(echo $(Build.SourceVersionMessage) | awk '{print $2}')
+              SHA_Before=$(echo $(Build.SourceVersionMessage) | awk '{print $4}')
+
+              python3 -m venv VENV
+              source VENV/bin/activate
+              cd $(WorkDir)/PerfX
+              pip3 install -q -r requirements.txt
+              module load /hpc/local/etc/modulefiles/hpcx-gcc
+
+              set -o pipefail; set -x
+              python3 -u perfx.py \
+                --config ucx-rdmz.yml \
+                --before "$SHA_Before" \
+                --after "$SHA_After" \
+                --module hpcx-gcc \
+                --filter \
+                --extra-ompi-opts " -H $(perf_nodes)" | tee $(WorkDir)/results.txt
+            displayName: Perf test
+            workingDirectory: $(WorkDir)
+
+          - script: $(WorkDir)/ucx/buildlib/tools/perf_results.py $(WorkDir)/results.txt $(threshold)
+            displayName: Results analyzer
+            workingDirectory: $(WorkDir)
+
+
+  - stage: Cleanup
```

Comment:
What happens artefacts with jobs that never reach cleanup stage? `Always` not always executed

Reply:
That's the purpose of `always`. Tested and it works.

Reply:
Not always
If pipeline fails as a pipeline you'll never reach this point and your artifacts will be there for ever

Reply:
How can we simulate such a situation?

Reply:
The cleanup stage is set with the agent's pool and no demands. It means that it will pick literally any agent available and send it cleaning.